src/Storages/MergeTree/MergeTreeReaderStream.cpp

```diff
@@ -195,18 +195,25 @@ void MergeTreeReaderStream::adjustForRange(MarkRange range)
     auto [right_offset, mark_range_bytes] = getRightOffsetAndBytesRange(range.begin, range.end);
     if (!right_offset)
     {
+        if (last_right_offset && !last_right_offset.value())
```

Comment:
Extremely confusing. `list_right_offset && list_right_offset == 0` right?

Reply:
Do we really need it? What happen if we just remove this `if`?

Reply:
Yes, we need it. Otherwise this will happen:
```
DB::Exception: Received from localhost:9000. DB::Exception: Prefetch is valid in readUntilEnd
```